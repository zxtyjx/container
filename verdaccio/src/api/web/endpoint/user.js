// @flow

import HTTPError from 'http-errors';
import type {Config} from '@verdaccio/types';
import type {Router} from 'express';
import type {IAuth, $ResponseExtend, $RequestExtend, $NextFunctionVer} from '../../../../types';

function addUserAuthApi(route: Router, auth: IAuth, config: Config) {
  route.post('/login', function(req: $RequestExtend, res: $ResponseExtend, next: $NextFunctionVer) {
    auth.authenticate(req.body.username, req.body.password, (err, user) => {
      if (!err) {
        req.remote_user = user;

        next({
          token: auth.issueUIjwt(user, '24h'),
          username: req.remote_user.name,
        });
      } else {
        next(HTTPError[err.message ? 401 : 500](err.message));
      }
    });
  });
}

export default addUserAuthApi;
