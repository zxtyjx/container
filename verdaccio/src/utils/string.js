// @flow

export function spliceURL(...args: Array<string>): string {
  return Array.from(args).reduce((lastResult, current) => lastResult + current).replace(/([^:])(\/)+(.)/g, `$1/$3`);
}
