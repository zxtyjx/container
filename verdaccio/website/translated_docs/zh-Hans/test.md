---
id: unit-testing(单元-测试）
title: "单元测试"
---
所有测试都被分成3 个文件夹：

- `test/unit` - 涵盖非平凡方式转换数据的功能测试。这些测试只 `require()` 一些文件并在其中运行代码，因此它们是非常快的。
- `test/functional` - 启动verdaccio instance并在 http上执行一系列请求的测试。它们比单元测试慢一些。
- `test/integration` - 启动verdaccio instance并用 npm对其执行请求的测试。它们真的很慢并能打击到真的npm registry。 **unmaintained test**

单元和功能测试是从项目根目录里运行 `npm test` 来自动执行的。集成测试应该时常手动执行。

我们所有测试都使用 `jest`。

## Npm 脚本

要运行测试脚本，您可以使用 `npm` 或 `yarn`。

    yarn run test
    

这将只会触发测试，单元和功能的前两组。

### 使用测试/单元

以下只是单元测试的一个例子。基本上遵守`jest` 标准。

请试着描述单元在 `test` 部分页眉里的单一句子里确切测试什么。

```javacript
const verdaccio = require('../../src/api/index');
const config = require('./partials/config');

describe('basic system test', () => {

  beforeAll(function(done) {
    // something important
  });

  afterAll((done) => {
    // undo something important
  });

  test('server should respond on /', done => {
    // your test
    done();
  });
});
```

### 使用测试/功能

Verdaccio 中的功能测试有点复杂，需要深入解释来让您有成功的体验。

一切从`index.js`文件开始。让我们来深入了解它吧。

```javascript
// we create 3 server instances
 const config1 = new VerdaccioConfig(
    './store/test-storage',
    './store/config-1.yaml',
    'http://localhost:55551/');
  const config2 = new VerdaccioConfig(
      './store/test-storage2',
      './store/config-2.yaml',
      'http://localhost:55552/');
  const config3 = new VerdaccioConfig(
        './store/test-storage3',
        './store/config-3.yaml',
        'http://localhost:55553/');
  const server1: IServerBridge = new Server(config1.domainPath);
  const server2: IServerBridge = new Server(config2.domainPath);
  const server3: IServerBridge = new Server(config3.domainPath);
  const process1: IServerProcess = new VerdaccioProcess(config1, server1, SILENCE_LOG);
  const process2: IServerProcess = new VerdaccioProcess(config2, server2, SILENCE_LOG);
  const process3: IServerProcess = new VerdaccioProcess(config3, server3, SILENCE_LOG);
  const express: any = new ExpressServer();
  ...

    // we check whether all instances has been started, since run in independent processes
    beforeAll((done) => {
      Promise.all([
        process1.init(),
        process2.init(),
        process3.init()]).then((forks) => {
          _.map(forks, (fork) => {
            processRunning.push(fork[0]);
          });
          express.start(EXPRESS_PORT).then((app) =>{
            done();
          }, (err) => {
            done(err);
          });
      }).catch((error) => {
        done(error);
      });
    });

    // after finish all, we ensure are been stoped
    afterAll(() => {
      _.map(processRunning, (fork) => {
        fork.stop();
      });
      express.server.close();
    });


```

### 使用

这里我们将描述常规功能测试看起来是什么样的，请核对内联了解更多详细信息。

#### The lib/server.js

服务器 class(类）只是模拟 `npm` client 的 wrapper类，它为功能测试提供简单的API。

如我们在之前的章节里提到的， 我们正创建3 个流程服务器，可以在每个流程里以`server1`, `server2` 和 ``server3`进行访问。

通过这样的引用，您可以给这任何3 个运行的instance 发送请求。

```javascript
<br />export default function(server) {
  // we recieve any server instance via arguments
  test('add tag - 404', () => {
    // we interact with the server instance.
    return server.addTag('testpkg-tag', 'tagtagtag', '0.0.1').status(404).body_error(/no such package/);
  });
});
```

### 测试/集成

这些部分还没有被使用，但是我们在寻求帮助来让它正常运转。**欢迎任何新的想法。**