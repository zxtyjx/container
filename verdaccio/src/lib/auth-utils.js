// @flow

import {ErrorCode} from './utils';
import {API_ERROR} from './constants';

import type {RemoteUser, Package, Callback} from '@verdaccio/types';

export function allow_action(action: string) {
  return function(user: RemoteUser, pkg: Package, callback: Callback) {
    const {name, groups} = user;
    const hasPermission = pkg[action].some((group) => name === group || groups.includes(group));

    if (hasPermission) {
      return callback(null, true);
    }

    if (name) {
      callback(ErrorCode.getForbidden(`user ${name} is not allowed to ${action} package ${pkg.name}`));
    } else {
      callback(ErrorCode.getForbidden(`unregistered users are not allowed to ${action} package ${pkg.name}`));
    }
  };
}

export function getDefaultPlugins() {
  return {
    authenticate(user: string, password: string, cb: Callback) {
      cb(ErrorCode.getForbidden(API_ERROR.BAD_USERNAME_PASSWORD));
    },

    add_user(user: string, password: string, cb: Callback) {
      return cb(ErrorCode.getConflict(API_ERROR.BAD_USERNAME_PASSWORD));
    },

    allow_access: allow_action('access'),
    allow_publish: allow_action('publish'),
  };
}
