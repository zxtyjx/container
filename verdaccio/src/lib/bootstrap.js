// @flow

import {assign, isObject, isFunction} from 'lodash';
import Path from 'path';
import URL from 'url';
import fs from 'fs';
import http from 'http';
import https from 'https';
// $FlowFixMe
import constants from 'constants';
import endPointAPI from '../api/index';
import {parse_address} from './utils';

import type {Callback} from '@verdaccio/types';
import type {$Application} from 'express';
import {DEFAULT_PORT} from './constants';

const logger = require('./logger');

/**
 * Retrieve all addresses defined in the config file.
 * Verdaccio is able to listen multiple ports
 * @param {String} argListen
 * @param {String} configListen
 * eg:
 *  listen:
    - localhost:5555
    - localhost:5557
    @return {Array}
 */
export function getListListenAddresses(argListen: string, configListen: mixed) {
  // command line || config file || default
  let addresses;
  if (argListen) {
    addresses = [argListen];
  } else if (Array.isArray(configListen)) {
    addresses = configListen;
  } else if (configListen) {
    addresses = [configListen];
  } else {
    addresses = [DEFAULT_PORT];
  }
  addresses = addresses.map(function(addr) {
    const parsedAddr = parse_address(addr);

    if (!parsedAddr) {
      logger.logger.warn({addr: addr},
         'invalid address - @{addr}, we expect a port (e.g. "4873"),'
       + ' host:port (e.g. "localhost:4873") or full url'
       + ' (e.g. "http://localhost:4873/")');
    }

    return parsedAddr;
  }).filter(Boolean);

  return addresses;
}

/**
 * Trigger the server after configuration has been loaded.
 * @param {Object} config
 * @param {Object} cliArguments
 * @param {String} configPath
 * @param {String} pkgVersion
 * @param {String} pkgName
 */
function startVerdaccio(config: any,
                        cliListen: string,
                        configPath: string,
                        pkgVersion: string,
                        pkgName: string,
                        callback: Callback) {
  if (isObject(config) === false) {
    throw new Error('config file must be an object');
  }

  endPointAPI(config).then((app)=> {
    const addresses = getListListenAddresses(cliListen, config.listen);

    addresses.forEach(function(addr) {
      let webServer;
      if (addr.proto === 'https') {
        // https  must either have key cert and ca  or a pfx and (optionally) a passphrase
        if (!config.https || !config.https.key || !config.https.cert || !config.https.ca) {
          displayHTTPSWarning(configPath);
        }

        webServer = handleHTTPS(app, configPath, config);
      } else { // http
        webServer = http.createServer(app);
      }

      unlinkAddressPath(addr);

      callback(webServer, addr, pkgName, pkgVersion);
    });
  });
}

function unlinkAddressPath(addr) {
  if (addr.path && fs.existsSync(addr.path)) {
    fs.unlinkSync(addr.path);
  }
}

function displayHTTPSWarning(storageLocation) {
  const resolveConfigPath = function(file) {
    return Path.resolve(Path.dirname(storageLocation), file);
  };

  logger.logger.fatal([
    'You have enabled HTTPS and need to specify either ',
    '    "https.key", "https.cert" and "https.ca" or ',
    '    "https.pfx" and optionally "https.passphrase" ',
    'to run https server',
    '',
    // commands are borrowed from node.js docs
    'To quickly create self-signed certificate, use:',
    ' $ openssl genrsa -out ' + resolveConfigPath('verdaccio-key.pem') + ' 2048',
    ' $ openssl req -new -sha256 -key ' + resolveConfigPath('verdaccio-key.pem') + ' -out ' + resolveConfigPath('verdaccio-csr.pem'),
    ' $ openssl x509 -req -in ' + resolveConfigPath('verdaccio-csr.pem') +
    ' -signkey ' + resolveConfigPath('verdaccio-key.pem') + ' -out ' + resolveConfigPath('verdaccio-cert.pem'),
    '',
    'And then add to config file (' + storageLocation + '):',
    '  https:',
    `    key: ${resolveConfigPath('verdaccio-key.pem')}`,
    `    cert: ${resolveConfigPath('verdaccio-cert.pem')}`,
    `    ca: ${resolveConfigPath('verdaccio-csr.pem')}`,
  ].join('\n'));
  process.exit(2);
}

function handleHTTPS(app, configPath, config) {
  try {
    let httpsOptions = {
      secureProtocol: 'SSLv23_method', // disable insecure SSLv2 and SSLv3
      secureOptions: constants.SSL_OP_NO_SSLv2 | constants.SSL_OP_NO_SSLv3,
    };

    if (config.https.pfx) {
      httpsOptions = assign(httpsOptions, {
        pfx: fs.readFileSync(config.https.pfx),
        passphrase: config.https.passphrase || '',
      });
    } else {
      httpsOptions = assign(httpsOptions, {
        key: fs.readFileSync(config.https.key),
        cert: fs.readFileSync(config.https.cert),
        ca: fs.readFileSync(config.https.ca),
      });
    }
    return https.createServer(httpsOptions, app);
  } catch (err) { // catch errors related to certificate loading
    logger.logger.fatal({err: err}, 'cannot create server: @{err.message}');
    process.exit(2);
  }
}

function listenDefaultCallback(webServer: $Application, addr: any, pkgName: string, pkgVersion: string) {
  webServer.listen(addr.port || addr.path, addr.host, () => {
    // send a message for tests
    if (isFunction(process.send)) {
      process.send({
        verdaccio_started: true,
      });
    }
  // $FlowFixMe
  }).on('error', function(err) {
    logger.logger.fatal({err: err}, 'cannot create server: @{err.message}');
    process.exit(2);
  });

  logger.logger.warn({
    addr: ( addr.path
          ? URL.format({
              protocol: 'unix',
              pathname: addr.path,
            })
          : URL.format({
              protocol: addr.proto,
              hostname: addr.host,
              port: addr.port,
              pathname: '/',
            })
          ),
    version: pkgName + '/' + pkgVersion,
  }, 'http address - @{addr} - @{version}');
}

export {startVerdaccio, listenDefaultCallback};
