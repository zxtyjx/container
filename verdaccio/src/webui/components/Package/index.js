import React from 'react';
import PropTypes from 'prop-types';
import {Tag} from 'element-react';
import {Link} from 'react-router-dom';

import {formatDateDistance} from '../../utils/package';

import classes from './package.scss';

const Package = ({name, version, author, description, license, time}) => {
  return (<section className={classes.package}>
    <Link to={`detail/${name}`}>
      <div className={classes.header}>
        <div className={classes.title}>
          <h1>
            {name} <Tag type="gray">v{version}</Tag>
          </h1>
        </div>
        <div role="author" className={classes.author}>
        { author ? `By: ${author}`: ''}
        </div>
      </div>
      <div className={classes.footer}>
        <p className={classes.description}>
          {description}
        </p>
      </div>
      <div className={classes.details}>
        <div className={classes.homepage}>
          {time ? `Published ${formatDateDistance(time)} ago` : ''}
        </div>
        <div className={classes.license}>
          {license}
        </div>
      </div>
    </Link>
  </section>);
};

Package.propTypes = {
  name: PropTypes.string,
  version: PropTypes.string,
  author: PropTypes.string,
  description: PropTypes.string,
  license: PropTypes.string,
  time: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Date)
  ])
};

export default Package;
