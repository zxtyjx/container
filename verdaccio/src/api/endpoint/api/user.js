// @flow

import type {$Response, Router} from 'express';
import type {$RequestExtend, $ResponseExtend, $NextFunctionVer, IAuth} from '../../../../types';
import {ErrorCode} from '../../../lib/utils';

import _ from 'lodash';
import Cookies from 'cookies';

export default function(route: Router, auth: IAuth) {
  route.get('/-/user/:org_couchdb_user', function(req: $RequestExtend, res: $Response, next: $NextFunctionVer) {
    res.status(200);
    next({
      ok: 'you are authenticated as "' + req.remote_user.name + '"',
    });
  });

  route.put('/-/user/:org_couchdb_user/:_rev?/:revision?', function(req: $RequestExtend, res: $Response, next: $NextFunctionVer) {
    let token = (req.body.name && req.body.password)
      ? auth.aesEncrypt(new Buffer(req.body.name + ':' + req.body.password)).toString('base64')
      : undefined;
    if (_.isNil(req.remote_user.name) === false) {
      res.status(201);
      return next({
        ok: 'you are authenticated as \'' + req.remote_user.name + '\'',
        token,
      });
    } else {
      auth.add_user(req.body.name, req.body.password, function(err, user) {
        if (err) {
          if (err.status >= 400 && err.status < 500) {
            // With npm registering is the same as logging in,
            // and npm accepts only an 409 error.
            // So, changing status code here.
            return next( ErrorCode.getCode(err.status, err.message) || ErrorCode.getConflict(err.message));
          }
          return next(err);
        }

        req.remote_user = user;
        res.status(201);
        return next({
          ok: 'user \'' + req.body.name + '\' created',
          token: token,
        });
      });
    }
  });

  route.delete('/-/user/token/*', function(req: $RequestExtend, res: $Response, next: $NextFunctionVer) {
    res.status(200);
    next({
      ok: 'Logged out',
    });
  });


  // placeholder 'cause npm require to be authenticated to publish
  // we do not do any real authentication yet
  route.post('/_session', Cookies.express(), function(req: $RequestExtend, res: $ResponseExtend, next: $NextFunctionVer) {
    res.cookies.set('AuthSession', String(Math.random()), {
      // npmjs.org sets 10h expire
      expires: new Date(Date.now() + 10 * 60 * 60 * 1000),
    });
    next({
      ok: true,
      name: 'somebody',
      roles: [],
    });
  });
}
