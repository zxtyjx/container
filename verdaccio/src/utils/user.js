// @flow
import {stringToMD5} from '../lib/crypto-utils';


export const GRAVATAR_DEFAULT = 'https://www.gravatar.com/avatar/00000000000000000000000000000000?d=mm';
/**
 * Generate gravatar url from email address
 */
export function generateGravatarUrl(email?: string): string {
  if (typeof email === 'string') {
    email = email.trim().toLocaleLowerCase();
    const emailMD5 = stringToMD5(email);

    return `https://www.gravatar.com/avatar/${emailMD5}`;
  } else {
    return GRAVATAR_DEFAULT;
  }
}
