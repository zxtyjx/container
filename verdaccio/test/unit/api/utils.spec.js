// @flow
import assert from 'assert';
import {generateGravatarUrl, GRAVATAR_DEFAULT}  from '../../../src/utils/user';
import {spliceURL}  from '../../../src/utils/string';
import Package from "../../../src/webui/components/Package/index";
import {validateName as validate, convertDistRemoteToLocalTarballUrls, parseReadme}  from '../../../src/lib/utils';
import Logger, {setup} from '../../../src/lib/logger';
import { readFile } from '../../functional/lib/test.utils';

const readmeFile = (fileName: string = 'markdown.md') => readFile(`../../unit/partials/readme/${fileName}`);

setup([]);

describe('Utilities', () => {

  describe('String utilities', () => {
    test('should splice two strings and generate a url', () => {
      const url: string = spliceURL('http://domain.com', '/-/static/logo.png');

      expect(url).toMatch('http://domain.com/-/static/logo.png');
    });

    test('should splice a empty strings and generate a url', () => {
      const url: string = spliceURL('', '/-/static/logo.png');

      expect(url).toMatch('/-/static/logo.png');
    });
  });

  describe('User utilities', () => {
      test('should generate gravatar url with email', () => {
        const gravatarUrl: string = generateGravatarUrl('user@verdaccio.org');

        expect(gravatarUrl).toMatch('https://www.gravatar.com/avatar/');
        expect(gravatarUrl).not.toMatch('000000000');
      });

      test('should generate generic gravatar url', () => {
        const gravatarUrl: string = generateGravatarUrl();

        expect(gravatarUrl).toMatch(GRAVATAR_DEFAULT);
      });
  });

  describe('Validations', () => {
    test('good ones', () => {
      assert( validate('verdaccio') );
      assert( validate('some.weird.package-zzz') );
      assert( validate('old-package@0.1.2.tgz') );
    });

    test('uppercase', () => {
      assert( validate('EVE') );
      assert( validate('JSONStream') );
    });

    test('no package.json', () => {
      assert( !validate('package.json') );
    });

    test('no path seps', () => {
      assert( !validate('some/thing') );
      assert( !validate('some\\thing') );
    });

    test('no hidden', () => {
      assert( !validate('.bin') );
    });

    test('no reserved', () => {
      assert( !validate('favicon.ico') );
      assert( !validate('node_modules') );
      assert( !validate('__proto__') );
    });

    test('other', () => {
      assert( !validate('pk g') );
      assert( !validate('pk\tg') );
      assert( !validate('pk%20g') );
      assert( !validate('pk+g') );
      assert( !validate('pk:g') );
    });
  });

  describe('Packages utilities', () => {
    const metadata: Package = {
      "name": "npm_test",
      "versions": {
        "1.0.0": {
          "dist": {
            "tarball": "http:\/\/registry.org\/npm_test\/-\/npm_test-1.0.0.tgz"
          }
        },
        "1.0.1": {
          "dist": {
            "tarball": "http:\/\/registry.org\/npm_test\/-\/npm_test-1.0.1.tgz"
          }
        }
      },
    };

    const buildURI = (host, version) => `http://${host}/npm_test/-/npm_test-${version}.tgz`;
    const host = 'fake.com';

    test('convertDistRemoteToLocalTarballUrls', () => {
      // $FlowFixMe
      const convertDist = convertDistRemoteToLocalTarballUrls(Object.assign({}, metadata), {
        headers: {
          host,
        },
        get: ()=> 'http',
        protocol: 'http'
      }, '');

      expect(convertDist.versions['1.0.0'].dist.tarball).toEqual(buildURI(host, '1.0.0'));
      expect(convertDist.versions['1.0.1'].dist.tarball).toEqual(buildURI(host, '1.0.1'));
    });
  });

  describe('parseReadme', () => {
    test('should pass for ascii/makrdown text to html template', () => {
      const markdown = '# markdown';
      const ascii = "= AsciiDoc";

      expect(parseReadme('testPackage', markdown)).toEqual('<h1 id="markdown">markdown</h1>\n');
      expect(parseReadme('testPackage', ascii)).toEqual('<h1>AsciiDoc</h1>\n');
      expect(parseReadme('testPackage', String(readmeFile('markdown.md')))).toMatchSnapshot();
      expect(parseReadme('testPackage', String(readmeFile('ascii.adoc')))).toMatchSnapshot();
    });

    test('should pass for conversion of non-ascii to markdown text', () => {
      const simpleText = 'simple text';
      const randomText = '%%%%%**##==';
      const randomTextNonAscii = 'simple text \n = ascii';
      const randomTextMarkdown = 'simple text \n # markdown';

      expect(parseReadme('testPackage', randomText)).toEqual('<p>%%%%%**##==</p>\n');
      expect(parseReadme('testPackage', simpleText)).toEqual('<p>simple text</p>\n');
      expect(parseReadme('testPackage', randomTextNonAscii))
        .toEqual('<p>simple text \n = ascii</p>\n');
      expect(parseReadme('testPackage', randomTextMarkdown))
        .toEqual('<p>simple text </p>\n<h1 id="markdown">markdown</h1>\n');
    });

    test('should show error for no readme data', () => {
      const noData = '';
      const spy = jest.spyOn(Logger.logger, 'error')
      expect(parseReadme('testPackage', noData))
        .toEqual('<p>ERROR: No README data found!</p>\n');
      expect(spy).toHaveBeenCalledWith({'packageName': 'testPackage'}, '@{packageName}: No readme found');
    });
  });
});
