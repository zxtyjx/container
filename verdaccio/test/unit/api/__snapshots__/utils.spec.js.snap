// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utilities parseReadme should pass for ascii/makrdown text to html template 1`] = `
"<h1 id=\\"project-title\\">Project Title</h1>
<p>One Paragraph of project description goes here</p>
<h2 id=\\"getting-started\\">Getting Started</h2>
<p>These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.</p>
<h3 id=\\"prerequisites\\">Prerequisites</h3>
<p>What things you need to install the software and how to install them</p>
<pre><code>Give examples
</code></pre><h3 id=\\"installing\\">Installing</h3>
<p>A step by step series of examples that tell you how to get a development env running</p>
<p>Say what the step will be</p>
<pre><code>Give the example
</code></pre><p>And repeat</p>
<pre><code>until finished
</code></pre><p>End with an example of getting some data out of the system or using it for a little demo</p>
<h2 id=\\"running-the-tests\\">Running the tests</h2>
<p>Explain how to run the automated tests for this system</p>
<h3 id=\\"break-down-into-end-to-end-tests\\">Break down into end to end tests</h3>
<p>Explain what these tests test and why</p>
<pre><code>Give an example
</code></pre><h3 id=\\"and-coding-style-tests\\">And coding style tests</h3>
<p>Explain what these tests test and why</p>
<pre><code>Give an example
</code></pre><h2 id=\\"deployment\\">Deployment</h2>
<p>Add additional notes about how to deploy this on a live system</p>
<h2 id=\\"built-with\\">Built With</h2>
<ul>
<li>The web framework used</li>
<li>Dependency Management</li>
<li>Used to generate RSS Feeds</li>
</ul>
<h2 id=\\"contributing\\">Contributing</h2>
<p>Please read <a href=\\"CONTRIBUTING.md\\">CONTRIBUTING.md</a>  for details on our code of conduct, and the process for submitting pull requests to us.</p>
<h2 id=\\"versioning\\">Versioning</h2>
<p>We use <a href=\\"http://semver.org/\\">SemVer</a> for versioning. For the versions available, see the <a href=\\"https://github.com/your/project/tags\\">tags on this repository</a>.</p>
<h2 id=\\"license\\">License</h2>
<p>This project is licensed under the MIT License - see the <a href=\\"LICENSE.md\\">LICENSE.md</a> file for details</p>
<h2 id=\\"acknowledgments\\">Acknowledgments</h2>
<ul>
<li>Hat tip to anyone whose code was used</li>
<li>Inspiration</li>
<li>etc</li>
</ul>
"
`;

exports[`Utilities parseReadme should pass for ascii/makrdown text to html template 2`] = `
"<h1>Hello, AsciiDoc!</h1>
<div id=\\"preamble\\">
<div class=\\"sectionbody\\">
<div class=\\"paragraph\\">
<p>An introduction to <a href=\\"http://asciidoc.org\\">AsciiDoc</a>.</p>
</div>
</div>
</div>
<div class=\\"sect1\\">
<h2 id=\\"_first_section\\">First Section</h2>
<div class=\\"sectionbody\\">
<div class=\\"ulist\\">
<ul>
<li>
<p>item 1</p>
</li>
<li>
<p>item 2</p>
</li>
</ul>
</div>
<div class=\\"listingblock\\">
<div class=\\"content\\">
<pre class=\\"highlight\\"><code class=\\"language-ruby\\" data-lang=\\"ruby\\">puts \\"Hello, World!\\"</code></pre>
</div>
</div>
</div>
</div>"
`;
